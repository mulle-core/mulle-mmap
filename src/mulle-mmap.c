/*
 * BASED ON:
 *
 * Copyright 2017 https://github.com/mandreyel
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this
 * software and associated documentation files (the "Software"), to deal in the Software
 * without restriction, including without limitation the rights to use, copy, modify,
 * merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies
 * or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
 * CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE
 * OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
#include "mulle-mmap.h"


#ifdef _WIN32
# include <windows.h>
#else
# include <unistd.h>
# include <fcntl.h>
# include <sys/mman.h>
# include <sys/stat.h>
#endif

#include "include-private.h"


uint32_t   mulle_mmap_get_version( void)
{
   return( MULLE_MMAP_VERSION);
}


/**
 * Determines the operating system's page allocation granularity.
 *
 * On the first call to this function, it invokes the operating system specific syscall
 * to determine the page size, caches the value, and returns it. Any subsequent call to
 * this function serves the cached value, so no further syscalls are made.
 */
static size_t  mulle_mmap_pagesize;


size_t  mulle_mmap_get_system_pagesize( void)
{
    if( ! mulle_mmap_pagesize)
    {
#ifdef _WIN32
        SYSTEM_INFO SystemInfo;
        GetSystemInfo(&SystemInfo);
        mulle_mmap_pagesize = SystemInfo.dwAllocationGranularity;
#else
        mulle_mmap_pagesize = sysconf(_SC_PAGE_SIZE);
#endif
    }
    return( mulle_mmap_pagesize);
}

/**
 * Aligns `offset` to the operating system page size such that it subtracts the
 * difference until the nearest lower page boundary before `offset`, or does
 * nothing if `offset` is already page aligned.
 */
static inline size_t   mulle_mmap_pagealign_offset( size_t offset)
{
    size_t page_size;

    page_size = mulle_mmap_get_system_pagesize();
    return( (offset / page_size) * page_size);
}


static inline void   *_mulle_mmap_get_mapping_start( struct mulle_mmap *p)
{
   char   *data;

   data = _mulle_mmap_get_data( p);
   if( data)
      data -= _mulle_mmap_get_mapping_offset( p);
   return( data);
}


#ifdef _WIN32

/** Returns the 4 upper bytes of an 8-byte integer. */
static inline DWORD   mulle_mmap_int64_high( int64_t n)
{
    return n >> 32;
}

/** Returns the 4 lower bytes of an 8-byte integer. */
static inline DWORD   mulle_mmap_int64_low( int64_t n)
{
    return n & 0xffffffff;
}

#endif // _WIN32




static mulle_mmap_file_t   mulle_mmap_file_open( char *path,
                                                 enum mulle_mmap_accessmode mode)
{
   mulle_mmap_file_t   handle;

#ifdef _WIN32
   handle = CreateFileA( path,
                         mode == mulle_mmap_read ? GENERIC_READ : (GENERIC_READ | GENERIC_WRITE),
                         FILE_SHARE_READ | FILE_SHARE_WRITE,
                         0,
                         OPEN_EXISTING,
                         FILE_ATTRIBUTE_NORMAL,
                         0);

#else // POSIX
    handle = open( path, mode == mulle_mmap_read ? O_RDONLY : O_RDWR);
#endif
    return( handle);
}


static inline int64_t   mulle_mmap_file_query_size( mulle_mmap_file_t handle)
{
#ifdef _WIN32
    LARGE_INTEGER   file_size;

    if( GetFileSizeEx( handle, &file_size) == 0)
        return ( -1);
	return( (int64_t) file_size.QuadPart);
#else // POSIX
    struct stat sbuf;

    if( fstat( handle, &sbuf) == -1)
      return( -1);
    return( sbuf.st_size);
#endif
}



struct memory_map_result
{
    char      *data;
    int64_t   length;
    int64_t   mapped_length;
#ifdef _WIN32
    mulle_mmap_file_t   file_mapping_handle;
#endif
};


static int   memory_map( mulle_mmap_file_t handle,
                         int64_t offset,
                         int64_t length,
                         enum mulle_mmap_accessmode mode,
                         struct memory_map_result *ctx)
{
    int64_t   aligned_offset;
    int64_t   length_to_map;
    char       *mapping_start;

    aligned_offset = mulle_mmap_pagealign_offset( offset);
    length_to_map  = offset - aligned_offset + length;

#ifdef _WIN32
   {
       int64_t            max_file_size;
       mulle_mmap_file_t   file_mapping_handle;

       max_file_size = offset + length;
       file_mapping_handle = CreateFileMapping(
               handle,
               0,
               mode == mulle_mmap_read ? PAGE_READONLY : PAGE_READWRITE,
               mulle_mmap_int64_high( max_file_size),
               mulle_mmap_int64_low( max_file_size),
               0);
       if( file_mapping_handle == MULLE_MMAP_INVALID_HANDLE)
         return( -1);

       mapping_start = (char *) MapViewOfFile(
               file_mapping_handle,
               mode == mulle_mmap_read ? FILE_MAP_READ : FILE_MAP_WRITE,
               mulle_mmap_int64_high( aligned_offset),
               mulle_mmap_int64_low( aligned_offset),
               length_to_map);
       if( mapping_start == nullptr)
         return( 1);
       ctx->file_mapping_handle = file_mapping_handle;
   }
#else // POSIX
    mapping_start = (char *) mmap(
            0, // Don't give hint as to where to map.
            length_to_map,
            mode == mulle_mmap_read ? PROT_READ : PROT_WRITE,
            MAP_SHARED,
            handle,
            aligned_offset);
    if( mapping_start == MAP_FAILED)
      return( -1);
#endif

    ctx->data          = mapping_start + offset - aligned_offset;
    ctx->length        = length;
    ctx->mapped_length = length_to_map;

    return( 0);
}


// -- mulle_mmap --

void  _mulle_mmap_done( struct mulle_mmap *p)
{
   _mulle_mmap_conditional_sync( p);
   _mulle_mmap_unmap( p);
}



int   _mulle_mmap_map_file_range( struct mulle_mmap *p,
                                  char *path,
                                  size_t offset,
                                  size_t length)
{
   mulle_mmap_file_t   handle;
   int                rval;

   handle = mulle_mmap_file_open( path, p->accessmode_);
   if( handle == MULLE_MMAP_INVALID_HANDLE)
      return( -1);

   rval = _mulle_mmap_map_range( p, handle, offset, length);
   if( ! rval)
      p->is_handle_internal_ = 1;

   return( rval);
}


int    _mulle_mmap_map_range( struct mulle_mmap *p,
                              mulle_mmap_file_t handle,
                              size_t offset,
                              size_t length)
{
   int64_t                    file_size;
   struct memory_map_result   ctx;
   int                        rval;

   file_size = mulle_mmap_file_query_size( handle);
   if( file_size == -1)
      return( -1);

    /*
     * quick check, though file size might have changed behind our back
     * already again
     */

    if( length != (size_t) -1 && offset + length > file_size)
    {
#ifdef _WIN32
#else
       errno = EINVAL;
#endif
       return( -1);
    }

    rval = memory_map( handle,
                       offset,
                       length == -1 ? (file_size - offset) : length,
                       p->accessmode_,
                       &ctx);
    if( ! rval)
    {
        // We must unmap the previous mapping that may have existed prior to this call.
        // Note that this must only be invoked after a new mapping has been created in
        // order to provide the strong guarantee that, should the new mapping fail, the
        // `map` function leaves this instance in a state as though the function had
        // never been invoked.
        _mulle_mmap_unmap( p);

        p->file_handle_         = handle;
        p->is_handle_internal_  = 0;
        p->data_                = ctx.data;
        p->length_              = ctx.length;
        p->mapped_length_       = ctx.mapped_length;
#ifdef _WIN32
        p->file_mapping_handle_ = ctx.file_mapping_handle;
#endif
    }
    return( rval);
}


int   _mulle_mmap_sync( struct mulle_mmap *p)
{
    if( ! _mulle_mmap_is_open( p))
        return( -1);

    if( _mulle_mmap_get_data( p))
    {
#ifdef _WIN32
        if( FlushViewOfFile( _mulle_mmap_get_mapping_start( p),
                             p->mapped_length_) == 0
           || FlushFileBuffers( p->file_handle_) == 0)
        {
           return( -1);
        }
#else // POSIX
        if( msync( _mulle_mmap_get_mapping_start( p), p->mapped_length_, MS_SYNC) != 0)
        {
            return( -1);
        }
#endif
    }

#ifdef _WIN32
    if( FlushFileBuffers( p->file_handle_) == 0)
       return( -1);
#endif
    return( 0);
}


int   _mulle_mmap_unmap( struct mulle_mmap *p)
{
   int   rval;

   rval = 0;
    if( ! _mulle_mmap_is_open( p))
        return( rval);

    // TODO do we care about errors here?
#ifdef _WIN32
    if( _mulle_mmap_is_mapped( p))
    {
        UnmapViewOfFile( _mulle_mmap_get_mapping_start( p));
        CloseHandle( p->file_mapping_handle_);
    }
#else // POSIX
    if( p->data_)
    {
       rval = munmap( _mulle_mmap_get_mapping_start( p), p->mapped_length_);
    }
#endif

    // If `file_handle_` was obtained by our opening it (when map is called with
    // a path, rather than an existing file handle), we need to close it,
    // otherwise it must not be closed as it may still be used outside this
    // instance.
    if( p->is_handle_internal_)
    {
#ifdef _WIN32
        CloseHandle( p->file_handle_);
#else // POSIX
        if( close( p->file_handle_))
           rval = -1;
#endif
    }

    // Reset fields to their default values.
    _mulle_mmap_init( p, p->accessmode_);
    return( rval);
}


int   _mulle_mmap_is_mapped( struct mulle_mmap *p)
{
#ifdef _WIN32
    return( p->file_mapping_handle_ != MULLE_MMAP_INVALID_HANDLE);
#else // POSIX
    return( _mulle_mmap_is_open( p));
#endif
}


int   _mulle_mmap_conditional_sync( struct mulle_mmap *p)
{
   if( p->accessmode_ == mulle_mmap_read)
      return( 0);
   return( _mulle_mmap_sync( p));
}

