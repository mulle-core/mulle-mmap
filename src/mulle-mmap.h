/*
 * BASED ON:
 *
 * Copyright 2017 https://github.com/mandreyel
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this
 * software and associated documentation files (the "Software"), to deal in the Software
 * without restriction, including without limitation the rights to use, copy, modify,
 * merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies
 * or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
 * CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE
 * OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
#ifndef mulle_mmap_h__
#define mulle_mmap_h__

#include "include.h"

#include <assert.h>
#include <errno.h>
#include <stdint.h>
#include <stddef.h>
#include <string.h>


#define MULLE_MMAP_VERSION  ((0 << 20) | (2 << 8) | 0)


static inline unsigned int   mulle_mmap_get_version_major( void)
{
   return( MULLE_MMAP_VERSION >> 20);
}


static inline unsigned int   mulle_mmap_get_version_minor( void)
{
   return( (MULLE_MMAP_VERSION >> 8) & 0xFFF);
}


static inline unsigned int   mulle_mmap_get_version_patch( void)
{
   return( MULLE_MMAP_VERSION & 0xFF);
}


MULLE_MMAP_GLOBAL
uint32_t   mulle_mmap_get_version( void);


#ifdef _WIN32
# ifndef WIN32_LEAN_AND_MEAN
#  define WIN32_LEAN_AND_MEAN
# endif // WIN32_LEAN_AND_MEAN
# include <windows.h>
#endif // ifdef _WIN32


/**
 * This is used by `mulle_mmap` to determine whether to create a read-only or
 * a read-write memory mapping.
 */


#ifdef _WIN32
typedef HANDLE  mulle_mmap_file_t;
# define MULLE_MMAP_INVALID_HANDLE  INVALID_HANDLE_VALUE
#else
typedef int     mulle_mmap_file_t;
# define MULLE_MMAP_INVALID_HANDLE  -1
#endif



enum mulle_mmap_accessmode
{
   mulle_mmap_read  = 0,
   mulle_mmap_write = 1
};


struct mulle_mmap
{
   void                         *data_;
   size_t                       length_;
   size_t                       mapped_length_;
   int                          is_handle_internal_;
   enum mulle_mmap_accessmode   accessmode_;
   /* end of memset 0 init */

   mulle_mmap_file_t            file_handle_;
#ifdef _WIN32
   mulle_mmap_file_t            file_mapping_handle_;
#endif
};


static inline void   _mulle_mmap_init( struct mulle_mmap *p,
                                       enum mulle_mmap_accessmode mode)
{
   memset( p, 0, (char *) &p->accessmode_ - (char *) p);

   p->accessmode_          = mode;
   p->file_handle_         = MULLE_MMAP_INVALID_HANDLE;
#ifdef _WIN32
   p->file_mapping_handle_ = MULLE_MMAP_INVALID_HANDLE;
#endif
}


/**
* If this is a read-write mapping, the destructor invokes sync. Regardless
* of the access mode, unmap is invoked as a final step.
*/
MULLE_MMAP_GLOBAL
void   _mulle_mmap_done( struct mulle_mmap *p);

/**
 * On UNIX systems 'file_handle' and 'mapping_handle' are the same. On Windows,
 * however, a mapped region of a file gets its own handle, which is returned by
 * 'mapping_handle'.
 */
static inline mulle_mmap_file_t
   _mulle_mmap_get_file_handle( struct mulle_mmap *p)
{
   return( p->file_handle_);
}


#ifdef _WIN32
static inline mulle_mmap_file_t
   _mulle_mmap_get_file_mapping_handle( struct mulle_mmap *p)
{
   return( p->file_mapping_handle_);
}
#endif


/*
 * A little additional interface to just get pages from the OS.
 */
//
// size should probably be a multiple of pagesize. It is known that the
// returned pages are zero filled!
//
MULLE_MMAP_GLOBAL
void   *mulle_mmap_alloc_pages( size_t size);
// free with this

MULLE_MMAP_GLOBAL
int     mulle_mmap_free_pages( void *p, size_t size);


/* like mulle_mmap_alloc_pages but produces shared memory pages instead
 * https://stackoverflow.com/questions/5656530/how-to-use-shared-memory-with-linux-in-c
 */
MULLE_MMAP_GLOBAL
void   *mulle_mmap_alloc_shared_pages( size_t size);

    /** Returns whether a valid memory mapping has been created. */
static inline int   _mulle_mmap_is_open( struct mulle_mmap *p)
{
   return( p->file_handle_ != MULLE_MMAP_INVALID_HANDLE);
}

    /**
     * Returns true if no mapping was established, that is, conceptually the
     * same as though the length that was mapped was 0. This function is
     * provided so that this class has Container semantics.
     */
static inline int   _mulle_mmap_is_empty( struct mulle_mmap *p)
{
   return( ! p->length_);
}


static inline int   _mulle_mmap_is_writable( struct mulle_mmap *p)
{
   return( p->accessmode_ == mulle_mmap_write);
}


MULLE_MMAP_GLOBAL
int   _mulle_mmap_is_mapped( struct mulle_mmap *p);

/**
 * `size` and `length` both return the logical length, i.e. the number of bytes
 * user requested to be mapped, while `mapped_length` returns the actual number of
 * bytes that were mapped which is a multiple of the underlying operating system's
 * page allocation granularity.
 */
static inline size_t   _mulle_mmap_get_length( struct mulle_mmap *p)
{
   return( p->length_);
}


static inline size_t   _mulle_mmap_get_mapped_length( struct mulle_mmap *p)
{
   return( p->mapped_length_);
}


static inline size_t   _mulle_mmap_get_mapping_offset( struct mulle_mmap *p)
{
   return( p->mapped_length_ - p->length_);
}


static inline void   *_mulle_mmap_get_data( struct mulle_mmap *p)
{
   return( p->data_);
}


/**
 * Establishes a memory mapping with AccessMode as was given during init.
 * If the mapping is unsuccessful, the return value is -1 and the struct
 * remains in a state as if this function hadn't been called.
 *
 * `path`, which must be a path to an existing file, is used to retrieve a file
 * handle (which is closed when the object destructs or `unmap` is called), which is
 * then used to memory map the requested region. Upon failure, `error` is set to
 * indicate the reason and the object remains in an unmapped state.
 *
 * `offset` is the number of bytes, relative to the start of the file, where the
 * mapping should begin. When specifying it, there is no need to worry about
 * providing a value that is aligned with the operating system's page allocation
 * granularity. This is adjusted by the implementation such that the first requested
 * byte (as returned by `data` or `begin`), so long as `offset` is valid, will be at
 * `offset` from the start of the file.
 *
 * `length` is the number of bytes to map. It may be `-1`, in which
 * case a mapping of the entire file is created.
 */
MULLE_MMAP_GLOBAL
int    _mulle_mmap_map_file_range( struct mulle_mmap *p,
                                   char *path,
                                   size_t offset,
                                   size_t length);

/**
 * Establishes a memory mapping with AccessMode as was given during init.
 * If the mapping is unsuccessful, the return value is -1 and the struct
 * remains in a state as if this function hadn't been called.
 *
 * `path`, which must be a path to an existing file, is used to retrieve a file
 * handle (which is closed when the object destructs or `unmap` is called), which is
 * then used to memory map the requested region. Upon failure, `error` is set to
 * indicate the reason and the object remains in an unmapped state.
 *
 * The entire file is mapped.
 */
static inline int   _mulle_mmap_map_file( struct mulle_mmap *p,
                                          char *path)
{
   return( _mulle_mmap_map_file_range( p, path, 0, (size_t) -1));
}

/**
 * Establishes a memory mapping with AccessMode as was given during init.
 * If the mapping is unsuccessful, the return value is -1 and the struct
 * remains in a state as if this function hadn't been called.
 *
 * `handle`, which must be a valid file handle, which is used to memory map the
 * requested region. Upon failure, `error` is set to indicate the reason and the
 * object remains in an unmapped state.
 *
 * `offset` is the number of bytes, relative to the start of the file, where the
 * mapping should begin. When specifying it, there is no need to worry about
 * providing a value that is aligned with the operating system's page allocation
 * granularity. This is adjusted by the implementation such that the first requested
 * byte (as returned by `data` or `begin`), so long as `offset` is valid, will be at
 * `offset` from the start of the file.
 *
 * `length` is the number of bytes to map. It may be `-1`, in which
 * case a mapping of the entire file is created.
 */
MULLE_MMAP_GLOBAL
int    _mulle_mmap_map_range( struct mulle_mmap *p,
                              mulle_mmap_file_t handle,
                              size_t offset,
                              size_t length);

/**
 * Establishes a memory mapping with AccessMode as was given during init.
 * If the mapping is unsuccessful, the return value is -1 and the struct
 * remains in a state as if this function hadn't been called.
 *
 * `handle`, which must be a valid file handle, which is used to memory map the
 * requested region. Upon failure, `error` is set to indicate the reason and the
 * object remains in an unmapped state.
 *
 * The entire file is mapped.
 */
static inline int  _mulle_mmap_map( struct mulle_mmap *p,
                                    mulle_mmap_file_t handle)
{
   return( _mulle_mmap_map_range( p, handle, 0, (size_t) -1));
}

/**
 * If a valid memory mapping has been created prior to this call, this call
 * instructs the kernel to unmap the memory region and disassociate this object
 * from the file.
 *
 * The file handle associated with the file that is mapped is only closed if the
 * mapping was created using a file path. If, on the other hand, an existing
 * file handle was used to create the mapping, the file handle is not closed.
 */
MULLE_MMAP_GLOBAL
int   _mulle_mmap_unmap( struct mulle_mmap *p);


/** Flushes the memory mapped page to disk. Errors are reported via `error`. */
MULLE_MMAP_GLOBAL
int   _mulle_mmap_sync( struct mulle_mmap *p);

/**
 * The destructor syncs changes to disk if `AccessMode` is `write`, but not
 * if it's `read`.
 */
MULLE_MMAP_GLOBAL
int   _mulle_mmap_conditional_sync( struct mulle_mmap *p);



static inline int   mulle_mmap_equal( struct mulle_mmap *p,
                                      struct mulle_mmap *q)
{
   return( p->accessmode_ == q->accessmode_ &&
           p->data_ == q->data_ &&
           p->length_ == q->length_);
}


MULLE_MMAP_GLOBAL
size_t  mulle_mmap_get_system_pagesize( void);


#ifdef __has_include
# if __has_include( "_mulle-mmap-versioncheck.h")
#  include "_mulle-mmap-versioncheck.h"
# endif
#endif

#endif
