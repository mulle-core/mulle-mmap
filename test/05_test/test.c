#include <mulle-mmap/mulle-mmap.h>

#include <stdio.h>
#include <stdlib.h>


static char  testfile[] = "test-file.dat";

void   test_at_offset( enum mulle_mmap_accessmode mode,
                       char *buffer,
                       size_t buflen,
                       char *filename,
                       size_t offset)
{
   struct mulle_mmap   info;
   size_t              mapped_size;
   char                *p;
   char                *sentinel;
   char                v;

   printf( "%s \"%s\" offset=%ld : ", mode ? "RW" : "R", filename, (long) offset);

    // Map the region of the file to which buffer was written.
   _mulle_mmap_init( &info, mode);
   if( _mulle_mmap_map_range_of_file( &info, filename, offset, -1))
      printf( "mmap failed %s\n", strerror( errno));

   if( ! _mulle_mmap_is_open( &info))
      printf( "not open %s\n", strerror( errno));

   mapped_size = buflen - offset;
   if( _mulle_mmap_get_length( &info) != mapped_size)
      printf( "buflen mismatch\n");

    // Start at first printable ASCII character.
    p         = &buffer[ 0];
    sentinel  = &buffer[ buflen];
    for( v = 33; p < sentinel; v = ++v < 126 ? v : 33)
      if( *p++ != v)
         printf( "contents mismatch\n");

   _mulle_mmap_done( &info);
   printf( "OK\n");
}




int main()
{
    size_t   page_size;
    size_t   file_size;
    char     *p;
    char     *buffer;
    char     *sentinel;
    char     v;
    FILE     *fp;
    int      mode;

    page_size = mulle_mmap_get_system_pagesize();
    if( ! page_size)
      return( 1);

    file_size = 4 * page_size - 250; // 16134, if page size is 4KiB
    buffer    = calloc( 1, file_size);

    // Start at first printable ASCII character.
    p         = &buffer[ 0];
    sentinel  = &buffer[ file_size];

    for( v = 33; p < sentinel; v = ++v < 126 ? v : 33)
        *p++ = v;

   fp = fopen( testfile, "w");
   if( ! fp)
      return( 2);

   fwrite( buffer, 1, file_size, fp);
   if( fclose( fp))
      return( 3);

   for( mode = 0; mode <= 1; mode++)
   {
      test_at_offset( mode, buffer, file_size, testfile, 0);
      test_at_offset( mode, buffer, file_size, testfile, page_size - 3);
      test_at_offset( mode, buffer, file_size, testfile, page_size + 3);
      test_at_offset( mode, buffer, file_size, testfile, 2 * page_size + 3);
      if( ! mode)
         printf( "\n");
   }

   return( 0);
}


