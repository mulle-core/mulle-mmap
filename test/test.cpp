#include <mio/mmap.hpp>
#include <mio/shared_mmap.hpp>

#include <string>
#include <fstream>
#include <cstdlib>
#include <iostream>
#include <cassert>
#include <system_error>
#include <numeric>

int handle_error(const std::error_code& error)
{
    const auto& errmsg = error.message();
    std::printf("error mapping file: %s, exiting...\n", errmsg.c_str());
    return error.value();
}

// Just make sure this compiles.
#ifdef CXX17
# include <cstddef>
using mmap_source = mio::basic_mmap_source<std::byte>;
#endif

int main()
{
    const char _path[] = "test-file";
    // Make sure mio compiles with non-const char* strings too.
    const int path_len = sizeof(_path);
    char* path = new char[path_len];
    std::copy(_path, _path + path_len, path);
    std::error_code error;
    // Fill buffer, then write it to file.
    const int file_size = 0x4000 - 250; // 16134
    std::string buffer(file_size, 0);
    // Start at first printable ASCII character.
    char v = 33;
    for (auto& b : buffer) {
       b = v;
       ++v; 
       // Limit to last printable ASCII character.
       v %= 126;
       if(v == 0) {
           v = 33;
       }
    }
    std::ofstream file(path);
    file << buffer;
    file.close();

    const size_t offset = 300;
    assert(offset < buffer.size());

    {
        // Map the region of the file to which buffer was written.
        mio::mmap_source file_view = mio::make_mmap_source(
                path, offset, mio::map_entire_file, error);
        if(error) { return handle_error(error); }

        const size_t mapped_size = buffer.size() - offset; // 15834
        assert(file_view.is_open());
        assert(file_view.size() == mapped_size);

        // Then verify that mmap's bytes correspond to that of buffer.
        for(size_t buf_idx = offset, view_idx = 0;
                buf_idx < buffer.size() && view_idx < mapped_size;
                ++buf_idx, ++view_idx) {
            if(file_view[view_idx] != buffer[buf_idx]) {
                std::printf("%luth byte mismatch: expected(%d) <> actual(%d)",
                        buf_idx, buffer[buf_idx], file_view[view_idx]);
                std::cout << std::flush;
                assert(0);
            }
        }

        // Turn file_view into a shared mmap.
        mio::shared_mmap_source shared_file_view(std::move(file_view));

        assert(!file_view.is_open());
        assert(shared_file_view.is_open());
        assert(shared_file_view.size() == mapped_size);

        // Then verify that shared_mmap's bytes correspond to that of buffer.
        for(size_t buf_idx = offset, view_idx = 0;
                buf_idx < buffer.size() && view_idx < mapped_size;
                ++buf_idx, ++view_idx) {
            if(shared_file_view[view_idx] != buffer[buf_idx]) {
                std::printf("%luth byte mismatch: expected(%d) <> actual(%d)",
                        buf_idx, buffer[buf_idx], shared_file_view[view_idx]);
                std::cout << std::flush;
                assert(0);
            }
        }
    }

    {
#define CHECK_INVALID_MMAP(m) do { \
        assert(error); \
        assert(m.empty()); \
        assert(!m.is_open()); \
        error.clear(); } while(0)

        mio::mmap_source m;

        // See if mapping an invalid file results in an error.
        m = mio::make_mmap_source("garbage-that-hopefully-doesnt-exist", 0, 0, error);
        CHECK_INVALID_MMAP(m);

        // Empty path?
        m = mio::make_mmap_source(static_cast<const char*>(0), 0, 0, error);
        CHECK_INVALID_MMAP(m);
        m = mio::make_mmap_source(std::string(), 0, 0, error);
        CHECK_INVALID_MMAP(m);

        // Invalid handle?
        m = mio::make_mmap_source(mio::invalid_handle, 0, 0, error);
        CHECK_INVALID_MMAP(m);

        // Invalid offset?
        m = mio::make_mmap_source(path, 100 * buffer.size(), buffer.size(), error);
        CHECK_INVALID_MMAP(m);
    }

    {
        // Make sure these compile.
        mio::ummap_source _1;
        mio::shared_ummap_source _2;
        // Make sure shared_mmap mapping compiles as all testing was done on
        // normal mmaps.
        mio::shared_mmap_source _3(path, 0, mio::map_entire_file);
        auto _4 = mio::make_mmap_source(path, error);
        auto _5 = mio::make_mmap<mio::shared_mmap_source>(path, 0, mio::map_entire_file, error);
#ifdef _WIN32
        const wchar_t* wpath1 = L"dasfsf";
        auto _6 = mio::make_mmap_source(wpath1, error);
        mio::mmap_source _7;
        _7.map(wpath1, error);
        const std::wstring wpath2 = wpath1;
        auto _8 = mio::make_mmap_source(wpath2, error);
        mio::mmap_source _9;
        _9.map(wpath1, error);
#endif
    }

    std::printf("all tests passed!\n");
}
