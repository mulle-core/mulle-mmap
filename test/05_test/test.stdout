R "test-file.dat" offset=0 : OK
R "test-file.dat" offset=4093 : OK
R "test-file.dat" offset=4099 : OK
R "test-file.dat" offset=8195 : OK

RW "test-file.dat" offset=0 : OK
RW "test-file.dat" offset=4093 : OK
RW "test-file.dat" offset=4099 : OK
RW "test-file.dat" offset=8195 : OK
